consensus/src/quorum_store/batch_generator.rs

```diff
@@ -95,9 +99,119 @@ impl BatchGenerator {
+        expiry_time: u64,
+        bucket_start: u64,
+    ) -> Batch {
+        let batch_id = self.batch_id;
+        self.batch_id.increment();
+        self.db
+            .save_batch_id(self.epoch, self.batch_id)
+            .expect("Could not save to db");
+
+        let txns_in_progress: Vec<_> = txns
+            .iter()
+            .map(|txn| TransactionInProgress {
+                summary: TransactionSummary {
+                    sender: txn.sender(),
+                    sequence_number: txn.sequence_number(),
+                },
+                gas_unit_price: txn.gas_unit_price(),
+            })
+            .collect();
+        self.batches_in_progress.insert(batch_id, txns_in_progress);
+        self.batch_expirations.add_item(batch_id, expiry_time);
+
+        counters::CREATED_BATCHES_COUNT.inc();
+        counters::num_txn_per_batch(bucket_start.to_string().as_str(), txns.len());
+
+        Batch::new(
+            batch_id,
+            txns,
+            self.epoch,
+            expiry_time,
+            self.my_peer_id,
+            bucket_start,
+        )
+    }
+
+    /// Push num_txns from txns into batches. If num_txns is larger than max size, then multiple
+    /// batches are pushed.
+    fn push_bucket_to_batches(
+        &mut self,
+        batches: &mut Vec<Batch>,
+        txns: &mut Vec<SignedTransaction>,
+        num_txns_in_bucket: usize,
+        expiry_time: u64,
+        bucket_start: u64,
+    ) -> bool {
+        let mut remaining_txns = num_txns_in_bucket;
+        while remaining_txns > 0 {
+            if batches.len() == self.config.sender_max_num_batches {
+                return false;
+            }
+            let num_batch_txns = std::cmp::min(self.config.sender_max_batch_txns, remaining_txns);
+            let batch_txns: Vec<_> = txns.drain(0..num_batch_txns).collect();
+            let batch = self.create_new_batch(batch_txns, expiry_time, bucket_start);
+            batches.push(batch);
+            remaining_txns -= num_batch_txns;
+        }
+        true
+    }
+
+    fn bucket_into_batches(
+        &mut self,
+        pulled_txns: &mut Vec<SignedTransaction>,
+        expiry_time: u64,
+    ) -> Vec<Batch> {
+        pulled_txns.sort_by_key(|txn| txn.gas_unit_price());
+
+        let mut batches = vec![];
+        let buckets = self.config.batch_buckets.clone();
+        for (&bucket_start, &bucket_end) in buckets.iter().tuple_windows() {
+            if pulled_txns.is_empty() {
+                break;
+            }
+
+            let num_txns_in_bucket = match pulled_txns
+                .binary_search_by_key(&(bucket_end, PeerId::ZERO), |txn| {
+                    (txn.gas_unit_price(), txn.sender())
+                }) {
+                Ok(index) => index,
+                Err(index) => index,
+            };
+            if num_txns_in_bucket == 0 {
+                continue;
+            }
+
+            let batches_space_remaining = self.push_bucket_to_batches(
```

Comment:
this looks like we're creating low gas price batches first?

Reply:
I'm actually confused, after we removed the gas price normalization, do we still have the monotonic increasing gas guarantee in the pulled txn?